ci/e2e-tests/helper-functions.sh

```diff
@@ -0,0 +1,179 @@
+#   $3 - DPS resource id
+#   $4 - tags for DPS instance
+function createDpsLinkedHub {
+    az iot dps linked-hub create \
+        --resource-group "$AZURE_RESOURCE_GROUP_NAME" --dps-name "$1" \
+        --connection-string "$(
+            az iot hub connection-string show \
+                --resource-group "$AZURE_RESOURCE_GROUP_NAME" --hub-name "$2" \
+                --query 'connectionString' --output tsv
+        )" \
+        --location "$(
+            az iot hub show \
+                --resource-group "$AZURE_RESOURCE_GROUP_NAME" --name "$2" \
+                --query 'location' --output tsv
+        )"
+
+    # `az iot dps linked-hub create` deletes the tags on the DPS that
+    # were set by `az iot dps create` for some unknown reason, so we
+    # need to tag it again.
+    #
+    # A bug in the latest az CLI causes az resource tag to use an API
+    # version not supported by the cloud. For now, we will manually
+    # specify the API version.
+    #
+    # Ref: https://github.com/Azure/azure-cli/issues/20263
+    >/dev/null az resource tag \
+        --ids "$3" \
+        --tags "$4" \
+        --api-version 2020-03-01
+}
+
+# Sets up a DPS custom allocation policy. The logic in the custom allocation policy
+# assigns devices to a 'foo-devices' IoT hub if they have a payload containing
+# a 'modelId' field with 'foo' in it. 
+function setupCustomAllocationPolicy {
+        echo 'Creating an Azure Function for use as a DPS custom allocation policy...' >&2
+        
+        # Initialize function app project and function
+        func init DpsCustomAllocationFunctionProj --dotnet
+        pushd DpsCustomAllocationFunctionProj
+        func new --name "$dps_allocation_function_name" --template "HTTP trigger" --authlevel "anonymous" --language C# --force
+        
+        # Copy source code into functino app project
+        cp "$GITHUB_WORKSPACE/ci/e2e-tests/DpsCustomAllocation.cs" .
+
+        # Add build deps
+        dotnet add package Microsoft.Azure.Devices.Provisioning.Service -v 1.16.3
+        dotnet add package Microsoft.Azure.Devices.Shared -v 1.27.0
+
+        # Create storeage account needed by function app
+        sa_name="sa${GITHUB_RUN_ID}r${GITHUB_RUN_NUMBER}"
+        az storage account create \
+            --name $sa_name \
+            --location $AZURE_LOCATION \
+            --resource-group $AZURE_RESOURCE_GROUP_NAME \
+            --sku Standard_LRS \
+            --tags "suite_id=$suite_id"
+
+        # Create function app
+        az functionapp create \
+            --resource-group $AZURE_RESOURCE_GROUP_NAME \
+            --consumption-plan-location $AZURE_LOCATION \
+            --runtime dotnet \
+            --functions-version 3 \
+            --name "$dps_allocation_functionapp_name" \
+            --disable-app-insights \
+            --storage-account $sa_name \
+            --tags "suite_id=$suite_id"
+
+        # Publishing the app sometimes fails, so retry up to 3 times
+        set +e
```

Comment:
You don't need to `set +e`. Just do `if func azure ....; then`

Reply:
Ah, that's how you prevent it from exiting upon non-zero exit code. Thanks. Done.